source/opt/fold_spec_const_op_composite.cpp

```diff
@@ -0,0 +1,762 @@
+            modified = true;
+          }
+          const_val_to_id_[const_value.get()] = inst->result_id();
+          id_to_const_val_[inst->result_id()] = std::move(const_value);
+        }
+        break;
+      }
+      // For a Spec Constants defined with OpSpecConstantOp instruction, check
+      // if it only depends on Normal Constants. If so, the Spec Constant will
+      // be folded. The original Spec Constant defining instruction will be
+      // replaced by Normal Constant defining instructions, and the new Normal
+      // Constants will be added to id_to_const_val_ and const_val_to_id_ so
+      // that we can use the new Normal Constants when folding following Spec
+      // Constants.
+      case SpvOp::SpvOpSpecConstantOp:
+        modified |= ProcessOpSpecConstantOp(&inst_iter);
+        break;
+      default:
+        break;
+    }
+  }
+  return modified;
+}
+
+bool FoldSpecConstantOpAndCompositePass::ProcessOpSpecConstantOp(
+    ir::Module::inst_iterator* pos) {
+  ir::Instruction* inst = &**pos;
+  ir::Instruction* folded_inst = nullptr;
+  assert(inst->GetInOperand(0).type ==
+             SPV_OPERAND_TYPE_SPEC_CONSTANT_OP_NUMBER &&
+         "The first in-operand of OpSpecContantOp instruction must be of "
+         "SPV_OPERAND_TYPE_SPEC_CONSTANT_OP_NUMBER type");
+
+  switch (static_cast<SpvOp>(inst->GetSingleWordInOperand(0))) {
+    case SpvOp::SpvOpCompositeExtract:
+      folded_inst = DoCompositeExtract(pos);
+      break;
+    case SpvOp::SpvOpVectorShuffle:
+      folded_inst = DoVectorShuffle(pos);
+      break;
+
+    case SpvOp::SpvOpCompositeInsert:
+      // Current Glslang does not generate code with OpSpecConstantOp
+      // CompositeInsert instruction, so this is not implmented so far.
+      // TODO(qining): Implement CompositeInsert case.
+      return false;
+
+    default:
+      // Component-wise operations.
+      folded_inst = DoComponentWiseOperation(pos);
+      break;
+  }
+  if (!folded_inst) return false;
+
+  // Replace the original constant with the new folded constant, kill the
+  // original constant.
+  uint32_t new_id = folded_inst->result_id();
+  uint32_t old_id = inst->result_id();
+  def_use_mgr_->ReplaceAllUsesWith(old_id, new_id);
+  def_use_mgr_->KillDef(old_id);
+  return true;
+}
+
+ir::Instruction* FoldSpecConstantOpAndCompositePass::DoCompositeExtract(
+    ir::Module::inst_iterator* pos) {
+  ir::Instruction* inst = &**pos;
+  assert(inst->NumInOperands() - 1 == 2 &&
+         "OpSpecConstantOp CompositeExtract requires two non-type non-opcode "
+         "operands.");
+  assert(inst->GetInOperand(1).type == SPV_OPERAND_TYPE_ID &&
+         "The vector operand must have a SPV_OPERAND_TYPE_ID type");
+  assert(
+      inst->GetInOperand(2).type == SPV_OPERAND_TYPE_LITERAL_INTEGER &&
+      "The literal operand must have a SPV_OPERAND_TYPE_LITERAL_INTEGER type");
+
+  analysis::Constant* first_operand_const =
+      FindRecordedConst(inst->GetSingleWordInOperand(1));
```

Comment:
Hmm, for OpSpecConstantOp, the first InOperand is the opcode: OpCompositeExtract. So I think the composite object is InOperand(1).